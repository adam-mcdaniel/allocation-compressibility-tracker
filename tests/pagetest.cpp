#include <stdio.h>
#include <stdlib.h>
#include <chrono>
#include <thread>
#include <unistd.h>
#include <sys/mman.h>
#include <ctype.h>

#define PAGE_SIZE 4096
#define SEED 0x12345678
#define u64 unsigned long long

// #define alloc(x) calloc(x, 1)
// #define alloc(x) malloc(x)
#define alloc(x) mmap(NULL, x, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)
#define free(x) munmap(x, 0)

int main() {
    srand(SEED);
    u64 number_of_pages = 16;
    while (number_of_pages < (16 << 5)) {
        u64 number_of_bytes = number_of_pages * PAGE_SIZE;

        char *ptr = (char*)alloc(number_of_bytes);
        // volatile char *ptr = (char*)mmap(NULL, number_of_bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
        // Use half of the allocated memory
        printf("Writing to %08llX (size=%lld)\n", (u64)ptr, number_of_bytes);
        for (int j=0; j<number_of_bytes / 4; j++) {
            ptr[j] = rand() % 256;
        }
        
        for (int j=0; j<number_of_bytes / 2; j++) {
            // printf("%c", ptr[j]);
            if (isalnum(ptr[j])) {
                printf("%X ", ptr[j]);
            } else if (ptr[j] == '\0') {
                printf(". ");
            } else {
                printf("? ");
            }
        }

        printf("Allocated %llu pages\n", number_of_pages);
        printf("Wrote to %llu bytes\n", number_of_bytes/2);

        // Sleep for 10 seconds
        std::this_thread::sleep_for(std::chrono::seconds(15));
        volatile int *x = (int*)alloc(1000);
        *x = 10;
        printf("x = %d\n", *x);
        free((void*)x);
        // free(ptr);

        number_of_pages <<= 1;
    }
    printf("Hello World!\n");
    return 0;
}