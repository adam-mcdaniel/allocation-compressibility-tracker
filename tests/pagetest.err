==257651== Memcheck, a memory error detector
==257651== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==257651== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==257651== Command: ./tests/pagetest.exe tests/pagetest.in
==257651== 
==257651== 
==257651== Process terminating with default action of signal 2 (SIGINT)
==257651==    at 0x496F7FA: clock_nanosleep@@GLIBC_2.17 (clock_nanosleep.c:78)
==257651==    by 0x49746E6: nanosleep (nanosleep.c:25)
==257651==    by 0x10974C: void std::this_thread::sleep_for<long, std::ratio<1l, 1l> >(std::chrono::duration<long, std::ratio<1l, 1l> > const&) (this_thread_sleep.h:82)
==257651==    by 0x109434: main (pagetest.cpp:45)
==257651== 
==257651== HEAP SUMMARY:
==257651==     in use at exit: 94,320 bytes in 44 blocks
==257651==   total heap usage: 5,435 allocs, 5,391 frees, 113,825,753 bytes allocated
==257651== 
==257651== LEAK SUMMARY:
==257651==    definitely lost: 0 bytes in 0 blocks
==257651==    indirectly lost: 0 bytes in 0 blocks
==257651==      possibly lost: 2,304 bytes in 1 blocks
==257651==    still reachable: 92,016 bytes in 43 blocks
==257651==         suppressed: 0 bytes in 0 blocks
==257651== Rerun with --leak-check=full to see details of leaked memory
==257651== 
==257651== For lists of detected and suppressed errors, rerun with: -s
==257651== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
