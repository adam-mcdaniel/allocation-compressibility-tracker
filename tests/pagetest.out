Allocated 16 pages
Wrote to 32768 bytes
Protecting allocations
Done protecting allocations
Size in pages: 1
0x7feeaaa80000 | PFN: 0x0013c013, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Size in pages: 1
0x7feeab080000 | PFN: 0x00480610, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Size in pages: 1
0x7feeab280000 | PFN: 0x00318612, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Size in pages: 1
0x7fee55480000 | PFN: 0x00319bc1, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Size in pages: 1
0x7fee55680000 | PFN: 0x00019003, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Size in pages: 16
0x7fee55880000 | PFN: 0x00019771, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55881000 | PFN: 0x00019772, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55882000 | PFN: 0x00019773, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55883000 | PFN: 0x00019774, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55884000 | PFN: 0x00019775, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55885000 | PFN: 0x00019776, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55886000 | PFN: 0x00019777, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55887000 | PFN: 0x00019778, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55888000 | PFN: 0x00019779, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee55889000 | PFN: 0x0001977a, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588a000 | PFN: 0x0001977b, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588b000 | PFN: 0x0001977c, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588c000 | PFN: 0x0001977d, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588d000 | PFN: 0x0001977e, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588e000 | PFN: 0x0001977f, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
0x7fee5588f000 | PFN: 0x00019000, Soft-dirty: 1, File/shared: 0, Swapped: 0, Present: 1, Flags: 0x400005828        Up-to-date, LRU, MMAP, Anon, Swap Backed, 
Unprotecting allocations
Done unprotecting allocations
Done with pass
Bucket stats:
Bucket 0 (1 bytes)
  No entries
Bucket 1 (2 bytes)
  No entries
Bucket 2 (3 bytes)
  No entries
Bucket 3 (4 bytes)
  No entries
Bucket 4 (5 bytes)
  No entries
Bucket 5 (6 bytes)
  No entries
Bucket 6 (7 bytes)
  No entries
Bucket 7 (8 bytes)
  No entries
Bucket 8 (9 bytes)
  No entries
Bucket 9 (10 bytes)
  No entries
Bucket 10 (16 bytes)
  No entries
Bucket 11 (32 bytes)
  Total compressed size: 41
  Total uncompressed size: 32
  Number of entries: 1
  Compression ratio (lower is better): 1.28125
  Total portion of heap (uncompressed): 0.000457875
  Total portion of heap (compressed): 0.00234312
Bucket 12 (64 bytes)
  No entries
Bucket 13 (128 bytes)
  No entries
Bucket 14 (256 bytes)
  Total compressed size: 42
  Total uncompressed size: 224
  Number of entries: 1
  Compression ratio (lower is better): 0.1875
  Total portion of heap (uncompressed): 0.00320513
  Total portion of heap (compressed): 0.00240027
Bucket 15 (512 bytes)
  No entries
Bucket 16 (1024 bytes)
  No entries
Bucket 17 (2048 bytes)
  No entries
Bucket 18 (4096 bytes)
  Total compressed size: 506
  Total uncompressed size: 4096
  Number of entries: 1
  Compression ratio (lower is better): 0.123535
  Total portion of heap (uncompressed): 0.0586081
  Total portion of heap (compressed): 0.0289176
Bucket 19 (65536 bytes)
  Total compressed size: 16909
  Total uncompressed size: 65536
  Number of entries: 1
  Compression ratio (lower is better): 0.258011
  Total portion of heap (uncompressed): 0.937729
  Total portion of heap (compressed): 0.966339
Bucket 20 (131072 bytes)
  No entries
Bucket 21 (262144 bytes)
  No entries
Bucket 22 (524288 bytes)
  No entries
Bucket 23 (1048576 bytes)
  No entries
Bucket 24 (2097152 bytes)
  No entries
Bucket 25 (4194304 bytes)
  No entries
Bucket 26 (8388608 bytes)
  No entries
Bucket 27 (1073741824 bytes)
  No entries
Site stats:
  Site 370 (./tests/pagetest.exe:0) at address 0x561aa8505342
    File site_stats___tests_pagetest_exe_561aa8505342.csv does not exist
    Bucket 0 (1 bytes)
      No entries
    Bucket 1 (2 bytes)
      No entries
    Bucket 2 (3 bytes)
      No entries
    Bucket 3 (4 bytes)
      No entries
    Bucket 4 (5 bytes)
      No entries
    Bucket 5 (6 bytes)
      No entries
    Bucket 6 (7 bytes)
      No entries
    Bucket 7 (8 bytes)
      No entries
    Bucket 8 (9 bytes)
      No entries
    Bucket 9 (10 bytes)
      No entries
    Bucket 10 (16 bytes)
      No entries
    Bucket 11 (32 bytes)
      Total compressed size: 41
      Total uncompressed size: 32
      Number of entries: 1
      Compression ratio (lower is better): 1.28125
      Total portion of allocation site's data (uncompressed): 1
      Total portion of allocation site's data (compressed): 1
    Bucket 12 (64 bytes)
      No entries
    Bucket 13 (128 bytes)
      No entries
    Bucket 14 (256 bytes)
      No entries
    Bucket 15 (512 bytes)
      No entries
    Bucket 16 (1024 bytes)
      No entries
    Bucket 17 (2048 bytes)
      No entries
    Bucket 18 (4096 bytes)
      No entries
    Bucket 19 (65536 bytes)
      No entries
    Bucket 20 (131072 bytes)
      No entries
    Bucket 21 (262144 bytes)
      No entries
    Bucket 22 (524288 bytes)
      No entries
    Bucket 23 (1048576 bytes)
      No entries
    Bucket 24 (2097152 bytes)
      No entries
    Bucket 25 (4194304 bytes)
      No entries
    Bucket 26 (8388608 bytes)
      No entries
    Bucket 27 (1073741824 bytes)
      No entries
  Site 675 (./tests/pagetest.exe:0) at address 0x561aa8505293
    File site_stats___tests_pagetest_exe_561aa8505293.csv does not exist
    Bucket 0 (1 bytes)
      No entries
    Bucket 1 (2 bytes)
      No entries
    Bucket 2 (3 bytes)
      No entries
    Bucket 3 (4 bytes)
      No entries
    Bucket 4 (5 bytes)
      No entries
    Bucket 5 (6 bytes)
      No entries
    Bucket 6 (7 bytes)
      No entries
    Bucket 7 (8 bytes)
      No entries
    Bucket 8 (9 bytes)
      No entries
    Bucket 9 (10 bytes)
      No entries
    Bucket 10 (16 bytes)
      No entries
    Bucket 11 (32 bytes)
      No entries
    Bucket 12 (64 bytes)
      No entries
    Bucket 13 (128 bytes)
      No entries
    Bucket 14 (256 bytes)
      Total compressed size: 42
      Total uncompressed size: 224
      Number of entries: 1
      Compression ratio (lower is better): 0.1875
      Total portion of allocation site's data (uncompressed): 0.00340633
      Total portion of allocation site's data (compressed): 0.00247773
    Bucket 15 (512 bytes)
      No entries
    Bucket 16 (1024 bytes)
      No entries
    Bucket 17 (2048 bytes)
      No entries
    Bucket 18 (4096 bytes)
      No entries
    Bucket 19 (65536 bytes)
      Total compressed size: 16909
      Total uncompressed size: 65536
      Number of entries: 1
      Compression ratio (lower is better): 0.258011
      Total portion of allocation site's data (uncompressed): 0.996594
      Total portion of allocation site's data (compressed): 0.997522
    Bucket 20 (131072 bytes)
      No entries
    Bucket 21 (262144 bytes)
      No entries
    Bucket 22 (524288 bytes)
      No entries
    Bucket 23 (1048576 bytes)
      No entries
    Bucket 24 (2097152 bytes)
      No entries
    Bucket 25 (4194304 bytes)
      No entries
    Bucket 26 (8388608 bytes)
      No entries
    Bucket 27 (1073741824 bytes)
      No entries
  Site 708 (./tests/pagetest.exe:0) at address 0x561aa85052f4
    File site_stats___tests_pagetest_exe_561aa85052f4.csv does not exist
    Bucket 0 (1 bytes)
      No entries
    Bucket 1 (2 bytes)
      No entries
    Bucket 2 (3 bytes)
      No entries
    Bucket 3 (4 bytes)
      No entries
    Bucket 4 (5 bytes)
      No entries
    Bucket 5 (6 bytes)
      No entries
    Bucket 6 (7 bytes)
      No entries
    Bucket 7 (8 bytes)
      No entries
    Bucket 8 (9 bytes)
      No entries
    Bucket 9 (10 bytes)
      No entries
    Bucket 10 (16 bytes)
      No entries
    Bucket 11 (32 bytes)
      No entries
    Bucket 12 (64 bytes)
      No entries
    Bucket 13 (128 bytes)
      No entries
    Bucket 14 (256 bytes)
      No entries
    Bucket 15 (512 bytes)
      No entries
    Bucket 16 (1024 bytes)
      No entries
    Bucket 17 (2048 bytes)
      No entries
    Bucket 18 (4096 bytes)
      Total compressed size: 506
      Total uncompressed size: 4096
      Number of entries: 1
      Compression ratio (lower is better): 0.123535
      Total portion of allocation site's data (uncompressed): 1
      Total portion of allocation site's data (compressed): 1
    Bucket 19 (65536 bytes)
      No entries
    Bucket 20 (131072 bytes)
      No entries
    Bucket 21 (262144 bytes)
      No entries
    Bucket 22 (524288 bytes)
      No entries
    Bucket 23 (1048576 bytes)
      No entries
    Bucket 24 (2097152 bytes)
      No entries
    Bucket 25 (4194304 bytes)
      No entries
    Bucket 26 (8388608 bytes)
      No entries
    Bucket 27 (1073741824 bytes)
      No entries
